### ISSUE TITLE
Allow configuration of hosts on worker nodes using --add-host flag

### ISSUE THREAD

**Issue by ciroque** [kubernetes-sigs/kind#3282]
**What would you like to be added**:

In the Docker provision module there is [logic that assembles the arguments](https://github.com/kubernetes-sigs/kind/blob/main/pkg/cluster/internal/providers/docker/provision.go#L215-L260) for running Nodes. While there are many arguments that can be specified, one that is missing is the [--add-host](https://docs.docker.com/engine/reference/commandline/run/#add-host) flag.

The addition of this flag would be very beneficial to alleviate an issue we have happened across, described below.

The general idea would be to add a new block (`extra_hosts` ?) to the configuration that defines a single flag instance to be passed through to the Docker run command, thus this:

```
extra_hosts {
    alias: host.docker.internal
    name: 172.17.0.1
}
```

would translate to: `--add-host=host.docker.internal:172.17.0.1`

Of course naming is hard, so those names are just placeholders pending discussion.

I am willing to do the work, scheduling and timing T.B.D. if the community believes this is a worthwhile effort.

**Why is this needed**:

### Docker Networking and the magic host

When running Docker on Mac and Windows there exists a magic host entry (host.docker.internal) that can be used to reference other containers running in Docker.
This host is not available, by default, when running Docker in Linux. It is possible to add this host when running an image with the `--add-host=host.docker.internal:host-gateway` flag, e.g.:

```shell
### Note: the --add-host flag is required only when running Docker on Linux...

docker run -rm -dit --add-host:host.docker.internal:host-gateway
```

This is a pretty straight-forward way to ensure the DNS entry is available in Linux.

While there are a number of available options for configuring system nodes using Kind; the option to add hosts does not exist.

This means there is no straight-forward way to add the magic Docker DNS entry to the Kind Worker Nodes when running on Linux.

## Options

Three options were considered for solving this issue:

1. Manually add the DNS entry into the worker nodes `/etc/hosts` file;
2. Submit a PR for [Kind](https://kind.sigs.k8s.io/) that adds support for the `--add-hosts` flag;
3. Use CoreDNS hosts entries to inject the magic host;

### Spike the DNS entries

This was a super hacky way of solving the issue by running a command on each worker node that added an entry to the hosts file for `host.docker.internal`.

```shell
docker exec -it mara-worker bash -c "echo 172.17.0.1 host.docker.internal >> /etc/hosts"
docker exec -it mara-worker2 bash -c "echo 172.17.0.1 host.docker.internal >> /etc/hosts"
```

Note: The `172.17.0.1` magic IP address is the default IP for the Docker bridge network.

### Submit a PR to Kind

This is actually the most desirable solution. It is clean, it improves the quality of an existing project that is widely used and respected, and it avoids any hackery to the Kubernetes cluster.

### Update CoreDNS hosts

This involves editing a ConfigMap in the Cluster used by CoreDNS for configuration. CoreDNS allows configuration of plugins
in the ConfigMap, one of which is the _hosts_ plugin (see References).

This involves a manual step, and manual input which can be error-prone; however, the editing mechanism in k8s includes a validation
step that disallows invalid ConfigMaps from being applied. So there is a safety net in place.

The full process is described in the [filebeat README](../../logging/agents/filebeat/README.md#dns-linux-only), but the general
idea is:

1. Use `kubectl edit` to edit the CoreDNS ConfigMap;
2. Add the magic host entry to the Corefile section;
3. Save the ConfigMap (validation happens automatically);
4. Restart the CoreDNS Pods;

Once CoreDNS is restarted the entry can be confirmed as noted in the filebeat README.

## Workaround

We are presently using the third option, updating the CoreDNS Corefile with the record, but as noted, being able to do this in Kind would be ideal.

**Comment by BenTheElder** [comment]
> extra_hosts {
    alias: host.docker.internal
    name: 172.17.0.1
}

I think we should just correctly handle this case instead of adding another config surface

https://github.com/kubernetes-sigs/kind/blob/3610f606516ccaa88aa098465d8c13af70937050/images/base/files/usr/local/bin/entrypoint#L452

**Comment by ciroque** [comment]
> > extra_hosts {
> > alias: host.docker.internal
> > name: 172.17.0.1
> > }
> 
> I think we should just correctly handle this case instead of adding another config surface
> 
> https://github.com/kubernetes-sigs/kind/blob/3610f606516ccaa88aa098465d8c13af70937050/images/base/files/usr/local/bin/entrypoint#L452

Okay, cool, I will take a closer look next week and come back with a proposal around updating that, thanks!

**Comment by aaomidi** [comment]
> I think we should just correctly handle this case instead of adding another config surface

I think another config surface makes a bit of sense here since this is technically a configurable surface on Docker (?).

It's hard to anticipate every need going into the future IMO, and this definitely isn't something that should be on by default I think either.

**Comment by BenTheElder** [comment]
> I think another config surface makes a bit of sense here since this is technically a configurable surface on Docker (?).

From a maintainability point of view we do NOT support all possible docker options. The goal is to provide Kubernetes clusters, and every knob is another possible broken interaction that we need to test and support.

> It's hard to anticipate every need going into the future IMO, and this definitely isn't something that should be on by default I think either.

It's not supportable to cover every possible usage anyhow.
I don't see why this couldn't be on by default, docker containers already have special hostnames like this.

**Comment by aaomidi** [comment]
>  I don't see why this couldn't be on by default, docker containers already have special hostnames like this.

Security considerations tbh. This allows a container to talk to a service that's on the host, and that by default seems like it can open a can of worms.

**Comment by aaomidi** [comment]
> This is possible with or without the host name. The hostname just makes it slightly less complicated to discover the IP.

Yep, sorry I forgot to come back to this issue later. You're right this has no impact on security considerations from that perspective.

**Comment by cthtrifork** [comment]
Any plans for implementing this? we are manually editing the /etc/host file at the moment inside the node containers

**Comment by BenTheElder** [comment]
> Any plans for implementing this? we are manually editing the /etc/host file at the moment inside the node containers

No, if there were it would be discussed here in an issue first.

See above: https://github.com/kubernetes-sigs/kind/issues/3282#issuecomment-1611898023

**Comment by stek29** [comment]
@BenTheElder 
> One thing not discussed so far: podman and [#3429](https://github.com/kubernetes-sigs/kind/pull/3429) (nerctl)

podman seems to enable `host.containers.internal` and `host.docker.internal` by default, the `add-host` flag is not needed: https://github.com/containers/podman/blob/1e7f810f714240f5d68f92baa1ab39ee53a249f5/docs/source/markdown/options/add-host.md
> Instead of an IP address, the special flag *host-gateway* can be given. This
> resolves to an IP address the container can use to connect to the host. The
> IP address chosen depends on your network setup, thus there's no guarantee that
> Podman can determine the *host-gateway* address automatically, which will then
> cause Podman to fail with an error message. You can overwrite this IP address
> using the *host_containers_internal_ip* option in *containers.conf*.
>
> The *host-gateway* address is also used by Podman to automatically add the
> `host.containers.internal` and `host.docker.internal` hostnames to `/etc/hosts`.
> You can prevent that by either giving the **--no-hosts** option, or by setting
> *host_containers_internal_ip="none"* in *containers.conf*. If no *host-gateway*
> address was configured manually and Podman fails to determine the IP address
> automatically, Podman will silently skip adding these internal hostnames to
> `/etc/hosts`.

so **for podman it's better not to pass** `--add-host host.docker.internal:host-gateway` since it might break if the `host-gateway` address is not known -- and if it's known it would've been added already, so using add-host makes no sense here.

---
nerdctl supports `host-gateway`, but it requires the `host-gateway` to be configured explicitly
https://github.com/containerd/nerdctl/blob/c01d0fa5747d38f8d5d445fb3d0909931221beaf/pkg/containerutil/containerutil.go#L620-L649

I'd say that **for nerdctl it's better not to pass** `--add-host host.docker.internal:host-gateway`, since it will break unless user has configured `host-gateway` IP explicitly.

---
docker on linux replaces the `host-gateway` with a `host-gateway-ip` / `host-gateway-ips` config, and if it's not provided -- it detects one as host-side address for the default bridge network:
https://github.com/moby/moby/blob/0de01f700d9040f70d71d16fff2ea4aafc4af36b/daemon/daemon_unix.go#L904-L920

so from what I see -- **for docker it should be safe to always pass** `--add-host host.docker.internal:host-gateway`.

for docker running in colima on macOS the `host.docker.internal` is not added by default – and it's basically the same as running docker on linux – except that colima sets the IP explicitly:
https://github.com/abiosoft/colima/blob/e79883bde6581b8ab0db8d0c63103cf4b1e89f8a/environment/container/docker/daemon.go#L13-L30

and as far as I know, with docker desktop it "just works" -- it's somewhat equivalent to having `--add-host host.docker.internal:host-gateway` always present

I guess they just have a special DNS server running for containers, which resolves that internal address

---
orbstack seems to resolve `host.docker.internal` by default from what I see in the docs:
https://docs.orbstack.dev/docker/network#connecting-to-servers-on-mac
and it's safe to pass `--add-host host.docker.internal:host-gateway` to it – it would just work too.
